cpp/include/raft/sparse/neighbors/detail/connect_components.cuh

```diff
@@ -182,48 +194,149 @@ struct LookupColorOp {
  * the given array of components
  * @tparam value_idx
  * @tparam value_t
+ * @param[in] handle raft handle
  * @param[out] kvp mapping of closest neighbor vertex and distance for each vertex in the given
  * array of components
  * @param[out] nn_colors components of nearest neighbors for each vertex
  * @param[in] colors components of each vertex
  * @param[in] X original dense data
  * @param[in] n_rows number of rows in original dense data
  * @param[in] n_cols number of columns in original dense data
- * @param[in] stream cuda stream for which to order cuda operations
+ * @param[in] row_batch_size row batch size for computing nearest neighbors
+ * @param[in] col_batch_size column batch size for sorting and 'unsorting'
+ * @param[in] reduction_op reduction operation for computing nearest neighbors
  */
 template <typename value_idx, typename value_t, typename red_op>
-void perform_1nn(raft::KeyValuePair<value_idx, value_t>* kvp,
+void perform_1nn(raft::resources const& handle,
+                 raft::KeyValuePair<value_idx, value_t>* kvp,
                  value_idx* nn_colors,
                  value_idx* colors,
                  const value_t* X,
                  size_t n_rows,
                  size_t n_cols,
-                 cudaStream_t stream,
+                 size_t row_batch_size,
+                 size_t col_batch_size,
                  red_op reduction_op)
 {
-  rmm::device_uvector<int> workspace(n_rows, stream);
-  rmm::device_uvector<value_t> x_norm(n_rows, stream);
-
-  raft::linalg::rowNorm(x_norm.data(), X, n_cols, n_rows, raft::linalg::L2Norm, true, stream);
-
-  raft::distance::fusedL2NN<value_t, raft::KeyValuePair<value_idx, value_t>, value_idx>(
-    kvp,
-    X,
-    X,
-    x_norm.data(),
-    x_norm.data(),
-    n_rows,
-    n_rows,
-    n_cols,
-    workspace.data(),
-    reduction_op,
-    reduction_op,
-    true,
-    true,
-    stream);
+  auto stream      = resource::get_cuda_stream(handle);
+  auto exec_policy = resource::get_thrust_policy(handle);
+
+  auto sort_plan = raft::make_device_vector<value_idx>(handle, (value_idx)n_rows);
+  raft::linalg::map_offset(handle, sort_plan.view(), [] __device__(value_idx idx) { return idx; });
+
+  thrust::sort_by_key(
+    resource::get_thrust_policy(handle), colors, colors + n_rows, sort_plan.data_handle());
+
+  // Modify the reduction operation based on the sort plan. This is particularly needed for HDBSCAN
```

Comment:
Please don't mention explicit use-cases downstream in docs for primitives. This layer should be more generalized.

Reply:
Done.